### ISSUE TITLE
In a multiple sites contest  all judgements may not be on all sites  (NPE in Contest XML report)

### ISSUE THREAD

**Issue by lane55** [pc2ccs/pc2v9#292]
**Describe the issue**: 

During a recent contest, the admin was not able to 

**To Reproduce**: 

Install pc2 version

unpack acpc-teens-2021-arch.zip

start server

show report, Report  ->  Contest XML Report

**Expected behavior**: 

   Create report with xml with no errors.

**Actual behavior**: please describe what _actually_ happened when you took the above steps:

NPE thrown

CSUS Programming Contest Control System
Version 9.7.0 20220106 (Thursday, January 6th 2022 04:19 UTC) Java ver 1.8.0_121 build 5822 Windows 10 10.0 (amd64) 
Contest Title: International Collegiate Programming Contest, Africa and Arab Collegiate Programming Championship for teens 2021
On: Friday, January 7, 2022 11:30:30 AM PST  Contest date/time: Sun Dec 05 01:00:34 PST 2021
Profile: Default (Default Contest)

** Contest XML Report

Exception in report: null
java.lang.NullPointerException
	at edu.csus.ecs.pc2.core.imports.ContestXML.addJudgementMemento(ContestXML.java:552)
	at edu.csus.ecs.pc2.core.imports.ContestXML.addRunMemento(ContestXML.java:522)
	at edu.csus.ecs.pc2.core.imports.ContestXML.toXML(ContestXML.java:168)
	at edu.csus.ecs.pc2.core.report.ContestReport.writeReport(ContestReport.java:44)
	at edu.csus.ecs.pc2.ui.ReportPane.createReportFile(ReportPane.java:534)
	at edu.csus.ecs.pc2.ui.ReportPane.generateSelectedReport(ReportPane.java:604)
	at edu.csus.ecs.pc2.ui.ReportPane$3.actionPerformed(ReportPane.java:438)
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
	at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348)
	at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
	at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:252)
	

**Environment**: please complete the following information:

All env.
211204 214058.165|INFO|main|info|Version 9.7.0 20210317 (Wednesday, March 17th 2021 06:30 UTC) Java ver 11.0.11 build 6181 Linux 5.4.0-1030-aws (amd64) 

**Log Info**:  

N/A

**Screenshots**:  

N/A 

**Additional context**: 

none.

**Comment by lane55** [comment]
On site 1 - there was no matching Yes judgement for run 100077.
When the report code tried to find the judgement it got a null
and that causes the NPE.

judgement id from run 100077 : Yes-5948497216174287731 , is a Site 2 yes judgement.

The only yes judgement on the site 1 model/judgement list : Yes--553507506826703089

For the contest the NPE happened when attempting to fetch a Judgement class
for run 100077.  

See below for each sites' model Yes from judgement list.

In English (sorta), the judgement for judgment element id  Yes-5948497216174287731
was not in site 1's model.   Since the Judgement class was not in the
site 1 model, returns null, causing the NPE.

--

Grabbed the yes judgements from the Judgement  report, report 19, and
There are 2 Yes judgements on site 2, 3 and 4.  

I ran report 19 and then grep'd for the Yes judgements.

Ex.
$ pc2report --site 1 --contestPassword <PASS> 19 > site1.judgements.txt
$ pc2report --site 2 --contestPassword <PASS> 19 > site2.judgements.txt
where <PASS> replaced with contest password

--

Here are the judgements and their id's for the Yes judgements

site1.judgements.txt:  'Yes' acronym=AC id=Yes--553507506826703089

site2.judgements.txt:  'Yes' acronym=AC id=Yes--553507506826703089
site2.judgements.txt:  'Yes' acronym=AC id=Yes-5948497216174287731

site3.judgements.txt:  'Yes' acronym=AC id=Yes--3292207760958135988
site3.judgements.txt:  'Yes' acronym=AC id=Yes--7734277600857896811

site4.judgements.txt:  'Yes' acronym=AC id=Yes--3292207760958135988
site4.judgements.txt:  'Yes' acronym=AC id=Yes--8730092025650960818

**Comment by lane55** [comment]
Requirement not met: all judgement lists in the contest should be present on all servers model

   This bug failed that requirement.

Requirement Site 1 judgements are authoritative after a remote site joins the contest.

   This may already be a met requirement, should test.

Up for discussion

Maybe new Requirement:  Disable judgement editing on other sites when in a connected contest.  
Enable judgement edit if the site is in not connected?

**Comment by clevengr** [comment]
Some additional data that may (or may not) be useful:
- There were 312 total runs in the contest (ACPC-Teens)
- There were 2 Judge accounts at Site 1 (J9 and J10); there was one judge account at Site 2 (J1)
- Site 1 shows the following "Judge" names in the "Judge" column on the "Runs" grid:
  - J9 S1 (39 runs)
  - J9 S1/AJ (95 runs)
  - J10 S1 (61 runs)
  - J10 S1/AJ (111 runs)
  - J1 S2 (3 runs)
  - J1 S2/AJ (3 runs)
- The **Configure Contest>AutoJudge** pane lists all three of the above judge accounts, BUT:
  - NONE of them show as "AutoJudge ON"
  - NONE of them show ANY "Problems selected"
- ACPC has historically NEVER used "Auto-Judging".
It's hard to understand why over half the runs in the contest (209 out of 312) are marked with "/AJ" on the Runs grid...

Additional info:  I put some debug statements in the code which have the effect of checking for "corrupted" runs (such as runs that have a JudgementId that doesn't exist in the contest model), and omitting any such runs from the XML report.  The debug shows three such runs:
```
Found corrupted run in ResolverEventFeedXML.toXML()! 
   Run number: 100077  Time = 63  ElementId = Run--8503350164041009513  Submitter = team224  Prob = brimore--1925454324698951061  JudgementRecord = {ElementId:JudgementRecord--1595101302798884489, JudgementId:Yes-5948497216174287731,judgerClientId:judge1Site2}  JudgementId = Yes-5948497216174287731  JudgementId not found in contest model!
Found corrupted run in ResolverEventFeedXML.toXML()! 
   Run number: 100078  Time = 66  ElementId = Run--2011289665575593645  Submitter = team218  Prob = holidays-83557830366780780  JudgementRecord = {ElementId:JudgementRecord--8711359991727369139, JudgementId:Yes-5948497216174287731,judgerClientId:judge1Site2}  JudgementId = Yes-5948497216174287731  JudgementId not found in contest model!
Found corrupted run in ResolverEventFeedXML.toXML()! 
   Run number: 100079  Time = 66  ElementId = Run-3908542588778542179  Submitter = team211  Prob = brimore--1925454324698951061  JudgementRecord = {ElementId:JudgementRecord-1021568949883010311, JudgementId:Yes-5948497216174287731,judgerClientId:judge1Site2}  JudgementId = Yes-5948497216174287731  JudgementId not found in contest model!
```

All three of these runs were judged by Judge1 on Site2 (and were the ONLY runs judged by that judge, with the exeception of those marked "/AJ", as noted above);  all three of these runs have the missing JudgementId `Yes-5948497216174287731`.  Removing those runs from the output produces an XML file that works correctly with the Resolver (V2.3.491, latest stable release as of 7 Jan 2022).  The final Resolver screen looks like this:


  
![ACPC-Teens ResolverResults](https://user-images.githubusercontent.com/9256691/148628918-7ccb57cb-cb2c-4973-9cd0-0532d82df2db.png)

**Comment by lane55** [comment]
The three runs confirm that the requirement from above
**Requirement not met: all judgement lists in the contest should be present on all servers' model**

**Comment by clevengr** [comment]
> The three runs confirm that the requirement from above **Requirement not met: all judgement lists in the contest should be present on all servers' models**

Right.  So how do we figure out why there was (at least) a Yes judgement on Site2 that was not present on Site1?

**Comment by lane55** [comment]
re: site 2 has site 1 runs.    Clearly Site 2 merged site 1 judgement list.   Not sure how that information adds anything new.
It is already known " all judgement lists in the contest should be present on all servers' models" and they aren't.

As to problem syncing, unknown and has not created a NPE.   While there is code that syncs judgements, problems, teams, contest times, etc.  If there is a different syncing problem that should be in a different syncing bug - because that can/should
be fixed/tested seperately.

**Comment by clevengr** [comment]
> re: site 2 has site 1 runs. Clearly Site 2 merged site 1 judgement list. Not sure how that information adds anything new. It is already known " all judgement lists in the contest should be present on all servers' models" and they aren't.

I'm not sure what you mean by `re: site 2 has site 1 runs.`  There is nothing in the discussion above about runs from one site being (or not being) present on another site.  

As far as the statement `Clearly Site 2 merged site 1 judgement list. Not sure how that information adds anything new. It is already known " all judgement lists in the contest should be present on all servers' models" and they aren't.`, yes, it's clear that the problem is that site 1 didn't get a judgement definition from site 2.   Stating that "X should happen and it didn't" doesn't help find out WHY.  Any data we can collect that MIGHT be relevant would seem to be useful to document (which is what I've been trying to do).

> 
> As to problem syncing, unknown and has not created a NPE. While there is code that syncs judgements, problems, teams, contest times, etc. If there is a different syncing problem that should be in a different syncing bug - because that can/should be fixed/tested seperately.

I would agree -- IF it could be documented that this unusual situation (unexpected indications of AutoJudging when no AutoJudging seems to have been selected) is not related to the issue of one site not having judgements from another site.  I would argue that one cannot know, a priori from just the facts we have so far, that these two things are unrelated.  That's why I noted this fact here.  If we later discover that they are indeed unrelated, then of course at that time we'll file a new, separate Issue.

**Comment by lane55** [comment]
Attached is my smoke test configuration for multi site testing.

In summary:
(legend:  sx -> primary  (site x logs into primary) 
s2 -> s1 
s3 -> s2
s4 -> s3
s5 -> s1

[multisite.smoke.test.info.config.txt](https://github.com/pc2ccs/pc2v9/files/7835478/multisite.smoke.test.info.config.txt)

**Comment by clevengr** [comment]
Did you look at whether the Judgement lists on the sites contain the same JudgementIds?  That would be useful to know.

Also, it would be useful to FIRST bring up Site 2 as "standalone" (so it generates its own Judgements), then shut it down and bring it back up as a remote site connecting to Site 1.  I suspect that may be what happened at ACPC; it would be useful to know if such a process would result in missing Judgements on one site or the other...

**Comment by lane55** [comment]
Yes.

Also, it would be useful to FIRST bring up Site 2 as "standalone".

I want to start with a smoke test and see if that reproduces the problem before changing the test.

**Comment by lane55** [comment]
Results:  Sites can connect into a contest normally (no startup standalone) and all judgements on all sites are not sync'd.

TEST
Using  site remote logins: s2 -> s1 ; s3 -> s2 ; s4 -> s3 ; s5 -> s1 (aka smoke test)
start sites
run report 19 (dump judgement lists) on all sites
compare all sites.

EXPECTED
All sites judgements found on all other sites, aka sync'd judgements.

Actual
Failed.
All sites do not have every other sites's judgments
single set of judgements on site 1, 9 judgements
Site 1, 2, 3, 4 all have Site 1's judgements
Site 5 does not have site 1's judgements (note s5 -> s1, even more strange)

Report 19 output for all sites
[smoketest.report.19.txt.txt](https://github.com/pc2ccs/pc2v9/files/7835541/smoketest.report.19.txt.txt)

**Comment by lane55** [comment]
Part of the root cause for this lack of sync is that there are global contest settings
and site specific contest settings.   The site specific settings are where the site is authoritative.
Site-specific settings include
ContestTime
Judgements
Accounts

Accounts are clearly sync'd as is ContestTime
There is a bug and Judgements are not being sync'd across the contest (site models)

**Comment by lane55** [comment]
The requirements and design are in this comment: https://github.com/pc2ccs/pc2v9/issues/292#issuecomment-1007696417

**Comment by clevengr** [comment]
Ok, so you are saying that this issue can be reproduced without any remote site first being brought up "standalone" -- so that's good (in a sense :)  
 That brings two questions to mind:
- Can we reproduce it with JUST TWO SITES? (It would greatly simplify testing if we could reproduce it with a minimal configuration like just two sites, and we are at least fairly certain that's what happened at ACPC.)
- Why in the world would it have taken until now for this to happen?  (We know that ACPC, at least, has been using multi-site for several years; how come it hasn't happened in the past?
  -- Is it possible that this is the first time that anyone has ever done multi-site _and also judged from more than one site_?

**Comment by troy2914** [comment]
> Ok, so you are saying that this issue can be reproduced without any remote site first being brought up "standalone" -- so that's good (in a sense :) That brings two questions to mind:
> 
>     * Can we reproduce it with JUST TWO SITES? (It would greatly simplify testing if we could reproduce it with a minimal configuration like just two sites, and we are at least fairly certain that's what happened at ACPC.)
> 
>     * Why in the world would it have taken until now for this to happen?  (We know that ACPC, at least, has been using multi-site for several years; how come it hasn't happened in the past?)
>       -- Is it possible that this is the first time that anyone has ever done multi-site _and also judged from more than one site_?

I am sure we have not done cross site judging in many many years in pacnw contest.   
Is the proper answer that the judgements are synced from site1 to all sites?  Or that siteX gets all other sites Judgements?

**Comment by lane55** [comment]
> Can we reproduce it with JUST TWO SITES?
Yes.  Will include smoke test note in next comment
> Why in the world would it have taken until now for this to happen? 
Luck.

**Comment by lane55** [comment]
Quick test, s2 -> s1, test if judgements are different (each site has its own judgements)

Result: bug reproduced, all sites' judgements are different.

Prep
 install pc2v9  into two different directories (site1 and site2)
   use Version 9.7.0 20220106 (Thursday, January 6th 2022 04:19 UTC) build 5822

Test Instructions
 Start server 1 with problemflagtest sample/cdp, it has 5 sites
   start server with --load samps\contests\problemflagtest (or --load problemflagtest)
  (aka start bin\pc2server --login s --contestpassword contest  --load samps\contests\problemflagtest )

 update/add site 2 pc2v9.ini remote server to: remoteServer=localhost:50002
 start server 2: bin\pc2server --login s2

 Run report 19 (Judgements Report) on both sites 
  in site1/ : C:\test\current\bin\pc2report --site 1 --contestPassword contest 19
  in site2/ : C:\test\current\bin\pc2report --site 2 --contestPassword contest 19

Expected

   Reproduce this bug, each Yes judgement will be different.

Actual 

   All site report output for two sites: [smoketest.judgements.2sites.txt](https://github.com/pc2ccs/pc2v9/files/7841607/smoketest.judgements.2sites.txt)
   fyi - (a 5 site test) All site report output for five sites: 
[smoketest.judgements.5sites.0110.txt](https://github.com/pc2ccs/pc2v9/files/7841609/smoketest.judgements.5sites.0110.txt)

   Summary, all sites' judgements are different, thus:

Site# Judgement
1	'Yes/Accepted' acronym=AC id=Yes/Accepted--878998112548438898
2	'Yes' acronym=AC id=Yes--714348272859820438
3	'Yes' acronym=AC id=Yes--3701241795925089156
4	'Yes' acronym=AC id=Yes--564575552540605678
5	'Yes' acronym=AC id=Yes-3460308417441531583

**Comment by clevengr** [comment]
This is useful information, but I'm not sure it completely accurately reproduces the bug.  It shows that there are indeed (in this case, 5) different Id's for Yes judgements (one on each site).  However, I believe that's not precisely the issue which occurred.  It doesn't seem illegal that there would be a different Id on each site.  The question is, why are the Id's from each site not present on other sites?

That is, when a remote site connects, why does that remote site's collection of Id's not get put into the Judgement hash table for each other site?  For ACPC, there was a site (e.g. 2) on which a run was judged and was assigned a judgement whose Id was defined on that site (2).  That run was correctly transmitted to Site 1, in the judged state.  Attempting to generate an (XML) report on Site 1 caused Site 1 to attempt to find the JudgementId from Site 2 in the Site 1 judgement hashtable (and it wasn't there, hence the NPE).

So we need to investigate why judgement Id's from Site 2 (or whatever) are not being pulled into the judgement hashtable on Site 1 when site 2 connects to Site 1.

Or, we need to redesign things so that remote sites never assign their own judgement id's (which I think would probably be a bad idea).

**Comment by lane55** [comment]
More analysis of zips.   This may muddy the waters rather than
clarify the issue.

From the following samples.  This 
confuses it the issue because 2 different primary servers seem
to be used.

I highly suspect that the zip files are from different contests.
I identified the sites by the db.? directory in the zip.

site 1: Luxor-Teens-pc2archive.zip
site 2: luxor-teens-pc2archive.zip
Site 3: qatar-pc2archive.zip
site 4: benin-pc2archive.zip

remoteServer map
s2 -> 10.0.4.231:50002
s3 -> 10.0.3.202:50002
s4 -> 10.0.3.202:50002

site1 C:/test/acpc2021/acpc-teens-2021-arch/pc2v9.ini

site2 C:/test/acpc2021/luxor-teens-pc2archive/pc2v9.ini
   remoteServer=10.0.4.231:50002
site3 C:/test/acpc2021/qatar-pc2archive/pc2v9.ini
   remoteServer=10.0.3.202:50002
site4 C:/test/acpc2021/benin-pc2archive/pc2v9.ini
   remoteServer=10.0.3.202:50002

From Judgement reports, only site 2 has the site 1 judgements.

site1.judgements.txt:  'Yes' acronym=AC id=Yes--553507506826703089

site2.judgements.txt:  'Yes' acronym=AC id=Yes--553507506826703089  ; site 1
site2.judgements.txt:  'Yes' acronym=AC id=Yes-5948497216174287731

site3.judgements.txt:  'Yes' acronym=AC id=Yes--3292207760958135988 
site3.judgements.txt:  'Yes' acronym=AC id=Yes--7734277600857896811

site4.judgements.txt:  'Yes' acronym=AC id=Yes--3292207760958135988 ; site 3?
site4.judgements.txt:  'Yes' acronym=AC id=Yes--8730092025650960818

**Comment by lane55** [comment]
Tested start site 2 stand alone, then joining with site 1

Same results as the smoke test.   

An aside, I was mildly surprised that it was allowed, I guessed that we might
have a feature where that would not be allowed.   Each contest has a ContestId and I thought we
might compare those to avoid mixing contests.  If  site1.contestId != site2.contestId and not site2 first (remote) login then deny
Clearly that is not a feature.

btw in the smoke test all Contest ID's are identical
From the ACPC teen Contest the contest IDs for site 1, 3 and 4 were different, site 1 and site 2 were identical

**Comment by lane55** [comment]
Still unclear how ACPC site 2 has both site 1 and site 2 judgements.

Something odd about Contest Ids in ACPC, from Profiles report.
This suggests that there were 2 contests, more mudding of the waters.

(used grep -i -e Contest.Id: -e Profiles.Report -e site stuf.Profiles all.profiles.report.txt)

Profiles Report
  site number : 1
    Contest ID: Default-3887609265879230028
  site number : 1
Profiles Report
  site number : 4
    Contest ID: Default--4460637682493158335
  site number : 4
Profiles Report
  site number : 2
    Contest ID: Default-3887609265879230028
  site number : 0
  site number : 2
  site number : 0
Profiles Report
  site number : 3
    Contest ID: Default--4460637682493158335
  site number : 3

In the smoke test the contest ids were identical
grep -i -e Contest.Id: -e Profiles.Report -e site pc2.13.report.txt
Site Directory: C:\test\current
Profiles Report
  site number : 1
    Contest ID: Default-7763640509263748991
  site number : 1
Site Directory: C:\test\current\site2
Profiles Report
  site number : 2
    Contest ID: Default-7763640509263748991
  site number : 2
Site Directory: C:\test\current\site3
Profiles Report
  site number : 3
    Contest ID: Default-7763640509263748991
  site number : 3
Site Directory: C:\test\current\site4
Profiles Report
  site number : 4
    Contest ID: Default-7763640509263748991
  site number : 4
Site Directory: C:\test\current\site5
Profiles Report
  site number : 5
    Contest ID: Default-7763640509263748991
  site number : 5

**Comment by lane55** [comment]
> This is useful information, but I'm not sure it completely accurately reproduces the bug. It shows that there are indeed (in this case, 5) different Id's for Yes judgements (one on each site). However, I believe that's not precisely the issue which occurred. It doesn't seem illegal that there would be a different Id on each site. The question is, why are the Id's from each site not present on other sites?

Reproducing the bug may never happen.  I've tried a number of tests and the only condition presented
was that always site 2 has its own set of judgements, so that means the original NPE (site 2 judgement
not on site 1) is not only possible but likely.

> That is, when a remote site connects, why does that remote site's collection of Id's not get put into the Judgement hash table for each other site? For ACPC, there was a site (e.g. 2) on which a run was judged and was assigned a judgement whose Id was defined on that site (2). That run was correctly transmitted to Site 1, in the judged state. Attempting to generate an (XML) report on Site 1 caused Site 1 to attempt to find the JudgementId from Site 2 in the Site 1 judgement hashtable (and it wasn't there, hence the NPE).
 
> So we need to investigate why judgement Id's from Site 2 (or whatever) are not being pulled into the judgement hashtable on Site 1 when site 2 connects to Site 1.
 
> Or, we need to redesign things so that remote sites never assign their own judgement id's (which I think would probably be a bad idea).

Or we write new code to avoid the bug condition, per REQ 1 (all servers have all judgements in their individual models).

We should implement a way to use/pick judgements if a site becomes disconnected.   Do we use site 1, do we
allow them to add judgements to from thier own site (using Judgement.getSiteNumber()).

**Comment by lane55** [comment]
At this point, it is likely that to reproduce is:
start site 1, configure
connect site 2
disconnect site 2 (by dropping the connection, since there is no auto reconnect)
submit run at site 2
judge run 
reconnect to site 1
run Contest XML report on site 1.

**Comment by clevengr** [comment]
```
Something odd about Contest Ids in ACPC, from Profiles report.
This suggests that there were 2 contests, more mudding of the waters.
```
I pursued this with Coach.  There WERE two contests:  the "ACPC-Adults" contest (aka "The ACPC") and the "ACPC-Teens" contest.

The overall setup was as follows:

ACPC-Adults:  four sites:  AWS (site 1, primary), Luxor (site2), Benin, and Qatar (Sites 3 and 4 but it's not certain at the moment which was 3 and which was 4).

ACPC-Teens:  three sites:  AWS (Site 1, primary), Luxor (site2), Qatar.  No teams participated from the Qatar site; he's not 100% sure whether the server was ever even started (he doesn't think so, but not certain).

One possibility is that they made a mistake and cross-referenced site IPs (i.e., pointed a server that was supposed to be in the Teens contest to the Adult contest); that would certainly cause some chaos.  But I think that before we presume that, we should try to replicate the issue just using the (two) sites from the Teens contest...

**Comment by clevengr** [comment]
I just confirmed that the ACPC bug can be easily reproduced, in just a two-site contest and without using any ACPC data at all.  Steps to do this are similar to what Doug described earlier:

- Start a site1 server, loading a preconfigured contest (I used **--load samps/sumithello**).
- Start a site2 server, _but do not put a **remoteserver=** entry in its pc2v9.ini_ (or said another way, _leave the default **#remoteserver=** entry commented-out)_.  This will cause the site2 server to come up "stand-alone", and in particular it will cause site 2 to generate its own "judgementIds".
- Shut down the site 2 server.
- Edit the site 2 server's pc2v9.ini to contain a **remoteserver=** entry pointing to the site 1 server.
- On the site 1 server, use the "Sites" tab to add site 2 to the contest.
- Restart the Site 2 server (so it connects to the site 1 primary server).
- Start a team on site 1; submit a run.
- Start a judge on site 2; select and judge the run.
- On site 2, generate a "Contest XML" report (or any other report that generates XML, such as "ICPC Tools Event Reed").  This will generate a **_correct report_**.
- On site 1, generate the same XML report(s).  They will throw NPEs because the judgement ID which was put into the judged run from Site 2 is not present in the table of judgement IDs on site 1 (whereas, both the site 1 and site 2 judgement ids ARE present in site 2.

I would be willing to bet that this is what happened at ACPC: somebody brought up site 2 "standalone", then everything got restarted with site 2 connecting to site 1 -- but containing some extra judgement IDs that weren't distributed to site 1.

So the next question is: how do we prevent this from happening?

**Comment by lane55** [comment]
On login ensure that all judgements are on all server by:
1 – adding missing judgements into the model

On add/modify of any judgement,  ensure that new or edited judgment
Is updated on all sites.

The next part is to decide the requirement for which site’s judgements
are used by judges on each site.   With 2 sites there can be 2 Yes judgements,
and we should only show the judge 1 Yes judgement.

Unless proxy site is used, and even if that is used, it should work too.
Every site added to the contest must login to every other site so
A LOGIN_SUCCESS packet is processed.

**Comment by clevengr** [comment]
Following a conversation between Doug and John, here is (John's understanding of) the list of actual code changes to be made to address this issue:

- Add code to the method which processes a LOGIN_SUCCESS packet for a Server (indicating that this server has successfully logged into its upstream remote server).  What this added code will do is
  - generate a list of all of THIS SITE's judgements (including the local judgementId, the judgement name (display text) and the judgement acronym)
  - send a new packet named (something like) JUDGEMENT_LIST_UPDATE to EVERY SITE in the contest.  This new packet will contain the list of judgements for this site (that is, for the site which just got a LOGIN_SUCCESS packet).
- Add code to allow a server to handle a JUDGEMENT_LIST_UPDATE packet.  What this code will do is merge the list of judgements received in the packet into the hashtable of judgements being maintained on this site.  (The effect of this is that each site will have a list of all judgements in the contest, regardless of which site the judgements were originally defined on.)
- Add code to the routine which populates the "Select Judgement" dropdown list on the Judge client so that if filters the set of judgements displayed to the (human) judge so that the list includes ONLY the judgements from Site 1 (which is deemed to be "authoritative" for judgement ids).

It was agreed to defer issues such as "what happens if someone on one site edits a judgement's text", since these types of issues are not strictly part of the ACPC problem which generated this issue.  At some point if these types of additional issues are to be considered, this should be done under a separate (new) Issue.

**Comment by clevengr** [comment]
UPDATE:
I have found a way to reproduce the ACPC error without the (somewhat unsatisfactory) requirement of first attempting to bring up site 2 in "standalone mode" (i.e., with no **remoteServer=** entry in the pc2v9.ini).  All that is required is to start site 2, WITH a **remoteServer=<site1IP:port>** entry, but that _site 1 is not running_.  

When this happens, the site 2 server throws a "login failed" exception, **_but it still creates local (site 2) judgements_**.   Later, when site 1 is started and then site 2 is restarted (and successfully connects to site 1), site 2 ends up with duplicate judgements (the set from site 1 PLUS its own set). 

In looking through the ACPC logs (see attached), I'm pretty certain THIS is what happened to them.
[SERVER0@site1-0.log](https://github.com/pc2ccs/pc2v9/files/7876216/SERVER0%40site1-0.log)

I do also believe that the proposed code changes listed in the previous comment will address this problem (although it's not a complete solution to the multi-site-judgements issue...)

**Comment by lane55** [comment]
tldr; Each site creates its own judgements on startup (whether connected or not).

We had thought that only if a site (s2) starts up stand alone does it create its own judgements, that is not true-ish.

This test report with s2 -> s1, shows each site creates their own judgements on startup.
[issue292.test.report01.txt](https://github.com/pc2ccs/pc2v9/files/7878022/issue292.test.report01.txt)

The design where a site always creates its own judgements it better than a site starting up and having no judgements
which would require an admin to manually add the "default" judgements.

**Comment by lane55** [comment]
Requirement is that all judgements from all sites are on all servers.

Test  Case 292-1 is:

use acpc archives for site 1 and 2 

- login servers (1 and 2) 
- submit and judge a run on site 2 
- run Contest XML report on site 1
- run Judgements report on site 1
- run Judgements report on site 2

Expected results

Contest XML report completes with no NPE
Judgements reports site 1 shows 16 judgements, 8 per site 
Judgements reports site 2 shows 16 judgements, 8 per site

Both judgement reports in first section show site=N where
N is the server where judgement was shown

Both judgement reports in (new) second section shows each
sites' judgements (8 each site)

**Comment by lane55** [comment]
Another requirement coded is: When a judgement is created on a site the siteNumber field should be assigned
the current site.   Before this fix all siteNumbers were/are assigned zero. 

This requirement was added for testing purposes.  When determining the site number for a judgement
they were all zero, report changes have been coded to make it very easy to see which judgements were
created on which sites.   This requirement will be needed in order to fix issue #293

**Comment by lane55** [comment]
This issue may have been fixed in #293.

**Comment by lane55** [comment]
closing issue, merged on Jan 19
